[Package]: Pluto

### Package description

A dialect of Lua 5.4, which adds desirable syntax and features to Lua which otherwise wouldn't exist.

### Home page URL

https://plutolang.github.io/

### Source code URL

https://github.com/well-in-that-case/Pluto

### Packaging policy acknowledgement

- [X] I have read and understand the [Packaging Policy](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy).

### Additional information

It's already compilable for android just do `make PLAT=linux` and it works

https://github.com/2096779623/termux-packages/runs/8032557098?check_suite_focus=true

```shell
g++ -std=c++17 -O3 -flto -Wall -Wextra -DLUA_COMPAT_5_3 -DLUA_USE_LINUX   -I/data/data/com.termux/files/usr/include  -c -o linit.o linit.cpp
ar rcu libpluto.a lapi.o lcode.o lctype.o ldebug.o ldo.o ldump.o lfunc.o lgc.o llex.o lmem.o lobject.o lopcodes.o lparser.o lstate.o lstring.o ltable.o ltm.o lundump.o lvm.o lzio.o lauxlib.o lbaselib.o lcorolib.o ldblib.o liolib.o lmathlib.o loadlib.o loslib.o lstrlib.o ltablib.o lutf8lib.o linit.o 
ar: `u' modifier ignored since `D' is the default (see `U')
ranlib libpluto.a
g++ -std=c++17 -O3 -flto -Wall -Wextra -DLUA_COMPAT_5_3 -DLUA_USE_LINUX   -I/data/data/com.termux/files/usr/include  -c -o lua.o lua.cpp
g++ -std=c++17 -O3 -flto -o pluto   lua.o libpluto.a -lm -ldl 
ld: error: undefined symbol: main
>>> referenced by /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/Scrt1.o:(_start)
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:69: pluto] Error 1
make[2]: Leaving directory '/home/builder/.termux-build/pluto/src/src'
make[1]: Leaving directory '/home/builder/.termux-build/pluto/src/src'
make[1]: *** [Makefile:120: linux-noreadline] Error 2
make: *** [Makefile:55: linux] Error 2
```

This package doesn't seem to have good support for cross compile.

--- a/packages/plutolang/Makefile.patch
+++ b/packages/plutolang/Makefile.patch
@@ -1,0 +1,10 @@
+--- a/Makefile
++++ b/Makefile
+@@ -59,7 +59,6 @@
+ 	cd src && $(INSTALL_EXEC) $(TO_BIN) $(INSTALL_BIN)
+ 	cd src && $(INSTALL_DATA) $(TO_INC) $(INSTALL_INC)
+ 	cd src && $(INSTALL_DATA) $(TO_LIB) $(INSTALL_LIB)
+-	cd doc && $(INSTALL_DATA) $(TO_MAN) $(INSTALL_MAN)
+ 
+ uninstall:
+ 	cd src && cd $(INSTALL_BIN) && $(RM) $(TO_BIN)

--- a/packages/plutolang/build.sh
+++ b/packages/plutolang/build.sh
@@ -1,0 +1,21 @@
+TERMUX_PKG_HOMEPAGE=https://plutolang.github.io/
+TERMUX_PKG_DESCRIPTION="A superset of Lua 5.4, with unique features, optimizations, and improvements"
+TERMUX_PKG_LICENSE="MIT"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=0.4.4
+TERMUX_PKG_SRCURL=https://github.com/PlutoLang/Pluto/archive/refs/tags/${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=78cba6b443fec3979ab466aef6869feb92ae2b4415042b18f32d4208c48018ff
+TERMUX_PKG_DEPENDS="libc++"
+TERMUX_PKG_BUILD_IN_SRC=true
+TERMUX_PKG_NO_STATICSPLIT=true
+TERMUX_PKG_EXTRA_MAKE_ARGS="
+PLAT=linux
+INSTALL_TOP=$TERMUX_PREFIX
+INSTALL_INC=$TERMUX_PREFIX/include/pluto
+"
+
+termux_step_pre_configure() {
+	CXXFLAGS+=" -std=c++17"
+	export MYCFLAGS="$CXXFLAGS $CPPFLAGS"
+	export MYLDFLAGS="$LDFLAGS"
+}

--- a/packages/plutolang/src-Makefile.patch
+++ b/packages/plutolang/src-Makefile.patch
@@ -1,0 +1,22 @@
+--- a/src/Makefile
++++ b/src/Makefile
+@@ -6,7 +6,7 @@
+ # Your platform. See PLATS for possible values.
+ PLAT= mingw
+ 
+-CXX= g++ -std=c++17 -O3 -flto
++CXX?= g++ -std=c++17 -O3 -flto
+ CXXFLAGS= -Wall -Wextra -DLUA_COMPAT_5_3 $(SYSCFLAGS) $(MYCFLAGS)
+ LDFLAGS= $(SYSLDFLAGS) $(MYLDFLAGS)
+ LIBS= -lm $(SYSLIBS) $(MYLIBS)
+@@ -25,8 +25,8 @@
+ SYSLDFLAGS=
+ SYSLIBS=
+ 
+-MYCFLAGS=
+-MYLDFLAGS=
++MYCFLAGS?=
++MYLDFLAGS?=
+ MYLIBS=
+ MYOBJS=
+ 

--- a/packages/plutolang/src-loslib.cpp.patch
+++ b/packages/plutolang/src-loslib.cpp.patch
@@ -1,0 +1,11 @@
+--- a/src/loslib.cpp
++++ b/src/loslib.cpp
+@@ -118,7 +118,7 @@
+ #define LUA_TMPNAMBUFSIZE	32
+ 
+ #if !defined(LUA_TMPNAMTEMPLATE)
+-#define LUA_TMPNAMTEMPLATE	"/tmp/lua_XXXXXX"
++#define LUA_TMPNAMTEMPLATE	"@TERMUX_PREFIX@/tmp/lua_XXXXXX"
+ #endif
+ 
+ #define lua_tmpnam(b,e) { \

--- a/packages/plutolang/src-luaconf.h.patch
+++ b/packages/plutolang/src-luaconf.h.patch
@@ -1,0 +1,11 @@
+--- a/src/luaconf.h
++++ b/src/luaconf.h
+@@ -214,7 +214,7 @@
+ 
+ #else			/* }{ */
+ 
+-#define LUA_ROOT	"/usr/local/"
++#define LUA_ROOT	"@TERMUX_PREFIX@/"
+ #define LUA_LDIR	LUA_ROOT "share/lua/" LUA_VDIR "/"
+ #define LUA_CDIR	LUA_ROOT "lib/lua/" LUA_VDIR "/"
+ 

